File: test-fixtures/languages/dart/micro/auth_service.dart
import 'dart:io';
import 'package:sqflite/sqflite.dart';

late Database authDb;

int? verifyToken(String token) {
  var query = "SELECT user_id FROM tokens WHERE token = '$token'";
  // sink: SQL injection via string interpolation
  authDb.rawQuery(query);
  return 1;
}

void runAdminCommand(int userId, String cmd) {
  // sink: command injection — cmd is concatenated into the shell
  Process.runSync('sh', ['-c', 'notify-admin $cmd']);
}

File: test-fixtures/languages/csharp/micro/AuthService.cs
using System.Data.SqlClient;
using System.Diagnostics;

namespace Micro
{
    public class AuthService
    {
        private readonly string _connStr = "Server=localhost;Database=auth;";

        public string VerifyToken(string token)
        {
            using var conn = new SqlConnection(_connStr);
            conn.Open();
            var query = "SELECT user_id FROM tokens WHERE token = '" + token + "'";
            var cmd = new SqlCommand(query, conn); // sink: SQL injection
            var result = cmd.ExecuteScalar();
            return result?.ToString();
        }

        public void RunAdminCommand(string userId, string cmd)
        {
            if (userId != null)
            {
                Process.Start("notify-admin", cmd); // sink: command injection
            }
        }
    }
}
